File: prototype/aidos_proto/plan.py
"""The plan document format.

This module turns a plan document into plain data, and plain data back into a
plan document. It knows nothing about the store, the database, or the command
line. Every function here is pure.

Document shape
--------------

parse_plan returns one dictionary with four keys:

    {"frontmatter": str, "preamble": str, "phases": [...],
     "context_sections": [...]}

A phase holds:

    {"number": int, "title": str, "state": str, "tickets": [...]}

A ticket holds:

    {"id": str, "title": str, "body": str, "criteria": str,
     "claimed_state": str, "order": int}

A context section holds:

    {"heading": str, "text": str, "index": int}

The heading keeps the "##" prefix. The text keeps every inner line. The parser
removes only the blank lines at the start and at the end of the text.

The index says how many phases come before the section in the document. The
renderer uses the index to put each section back between the right phases. The
index is optional. The renderer puts a section with no index after the last
phase.

Text rules
----------

* Frontmatter is optional. It starts on line 1 with a line that holds only
  three hyphens. It ends on the next line that holds only three hyphens. The
  parser keeps the frontmatter text, and the two fence lines, without a change.
* The text between the frontmatter and the first "##" heading is the preamble.
  The parser keeps the preamble without a change.
* A phase heading has this form, with an em dash before the state:

      ## Phase 1: Groundwork - `done`

* A ticket line has this form:

      - [ ] **Ticket 1: Read the kernel.** A body. **Evaluate:** A criterion.

  The title carries no period. The renderer adds the period. The marker
  "**Evaluate:**" is required. It may sit on a continuation line.
* A continuation line starts with two spaces. It adds one line to the body of
  the ticket above it.
* Inside a phase, only a blank line, a ticket line, and a continuation line are
  legal. Any other line raises PlanParseError with its line number.
* Any other "##" heading starts a context section.

Round trip
----------

render_plan(parse_plan(text)) gives a document that parses to the same data.
Two renders of the same data give the same bytes. The renderer walks the
phases, the tickets, and the context sections in document order, and it sorts
nothing.
"""

import re

# The mark that each line carries, and the state that the mark claims.
MARK_STATES = {
    " ": "open",
    "~": "in_progress",
    "?": "awaiting_verification",
    "x": "done",
}

# The reverse map, which the renderer uses.
STATE_MARKS = {state: mark for mark, state in MARK_STATES.items()}

# The fence that opens and closes the frontmatter.
FENCE = "---"

# The prefix of every heading that opens a phase or a context section.
HEADING_PREFIX = "## "

# The prefix of a continuation line.
CONTINUATION_PREFIX = "  "

# The marker that separates the body of a ticket from its criteria.
CRITERIA_MARKER = "**Evaluate:**"

PHASE_HEADING = re.compile(
    r"^## Phase (\d+):\s*(.+?)\s*\u2014\s*`([^`]*)`\s*$")

TICKET_LINE = re.compile(
    r"^- \[([ ~?x])\] \*\*Ticket ([^:]+): (.+?)\.\*\*\s?(.*)$")


class PlanParseError(Exception):
    """One line of a plan document that the parser refuses.

    The error carries the 1-based number of the line and a message. The text
    of the error names the line number.
    """

    def __init__(self, line, message):
        super().__init__(message)
        self.line = line
        self.message = message

    def __str__(self):
        return "line %d: %s" % (self.line, self.message)


def parse_plan(text):
    """Read one plan document and return its data.

    The function raises PlanParseError on the first line that breaks a rule.
    """
    lines = text.split("\n")
    frontmatter, index = _take_frontmatter(lines)
    preamble, index = _take_preamble(lines, index)
    phases = []
    context_sections = []
    while index < len(lines):
        match = PHASE_HEADING.match(lines[index])
        if match:
            phase, index = _take_phase(match, lines, index)
            phases.append(phase)
        else:
            section, index = _take_context_section(lines, index, len(phases))
            context_sections.append(section)
    return {
        "frontmatter": frontmatter,
        "preamble": preamble,
        "phases": phases,
        "context_sections": context_sections,
    }


def render_plan(doc):
    """Write one plan document from the data that parse_plan returns."""
    blocks = []
    frontmatter = doc["frontmatter"].strip()
    if frontmatter:
        blocks.append(frontmatter)
    preamble = doc["preamble"].strip()
    if preamble:
        blocks.append(preamble)
    phases = doc["phases"]
    sections = doc["context_sections"]
    for position in range(len(phases) + 1):
        for section in sections:
            if _section_position(section, len(phases)) == position:
                blocks.append(_render_context_section(section))
        if position < len(phases):
            blocks.append(_render_phase(phases[position]))
    if not blocks:
        return ""
    return "\n\n".join(blocks) + "\n"


# ---- reading ----


def _take_frontmatter(lines):
    """Return the frontmatter text and the index of the line after it."""
    if not lines or lines[0].strip() != FENCE:
        return "", 0
    for index in range(1, len(lines)):
        if lines[index].strip() == FENCE:
            return "\n".join(lines[:index + 1]), index + 1
    raise PlanParseError(1, "the frontmatter never closes")


def _take_preamble(lines, start):
    """Return the preamble text and the index of the first heading."""
    index = start
    while index < len(lines) and not _is_heading(lines[index]):
        index += 1
    return _trim_blank_lines(lines[start:index]), index


def _take_phase(match, lines, start):
    """Return one phase and the index of the line that ends it."""
    raw_tickets = []
    index = start + 1
    while index < len(lines) and not _is_heading(lines[index]):
        line = lines[index]
        number = index + 1
        if line.startswith(CONTINUATION_PREFIX) and line.strip():
            if not raw_tickets:
                raise PlanParseError(
                    number, "line %d continues a ticket, but the phase holds "
                    "no ticket yet" % number)
            raw_tickets[-1]["lines"].append(line.strip())
        elif line.strip():
            ticket_match = TICKET_LINE.match(line)
            if not ticket_match:
                raise PlanParseError(
                    number, "line %d is neither a ticket line nor a "
                    "continuation line" % number)
            raw_tickets.append(
                _start_ticket(ticket_match, number, len(raw_tickets) + 1))
        index += 1
    phase = {
        "number": int(match.group(1)),
        "title": match.group(2),
        "state": match.group(3),
        "tickets": [_finish_ticket(raw) for raw in raw_tickets],
    }
    return phase, index


def _take_context_section(lines, start, phase_count):
    """Return one context section and the index of the line that ends it."""
    index = start + 1
    while index < len(lines) and not _is_heading(lines[index]):
        index += 1
    section = {
        "heading": lines[start].rstrip(),
        "text": _trim_blank_lines(lines[start + 1:index]),
        "index": phase_count,
    }
    return section, index


def _start_ticket(match, line_number, order):
    """Return the working data of one ticket, before its body is complete."""
    first = match.group(4).strip()
    return {
        "id": match.group(2),
        "title": match.group(3),
        "lines": [first] if first else [],
        "claimed_state": MARK_STATES[match.group(1)],
        "order": order,
        "line": line_number,
    }


def _finish_ticket(raw):
    """Split the body of one ticket from its criteria and return the ticket."""
    text = "\n".join(raw["lines"])
    if CRITERIA_MARKER not in text:
        raise PlanParseError(
            raw["line"], "line %d starts a ticket that holds no %s marker"
            % (raw["line"], CRITERIA_MARKER))
    body, _, criteria = text.partition(CRITERIA_MARKER)
    return {
        "id": raw["id"],
        "title": raw["title"],
        "body": body.strip(),
        "criteria": criteria.strip(),
        "claimed_state": raw["claimed_state"],
        "order": raw["order"],
    }


def _is_heading(line):
    """Say whether one line opens a phase or a context section."""
    return line.startswith(HEADING_PREFIX)


def _trim_blank_lines(block):
    """Join lines, and drop the blank lines at the start and at the end."""
    start = 0
    end = len(block)
    while start < end and not block[start].strip():
        start += 1
    while end > start and not block[end - 1].strip():
        end -= 1
    return "\n".join(block[start:end])


# ---- writing ----


def _section_position(section, phase_count):
    """Return the number of phases that come before one context section."""
    position = section.get("index", phase_count)
    return max(0, min(position, phase_count))


def _render_context_section(section):
    """Return the text of one context section, heading first."""
    text = section["text"].strip()
    if not text:
        return section["heading"]
    return "%s\n\n%s" % (section["heading"], text)


def _render_phase(phase):
    """Return the text of one phase, heading first, then every ticket."""
    lines = ["## Phase %d: %s \u2014 `%s`"
             % (phase["number"], phase["title"], phase["state"])]
    if phase["tickets"]:
        lines.append("")
        for ticket in phase["tickets"]:
            lines.extend(_render_ticket(ticket))
    return "\n".join(lines)


def _render_ticket(ticket):
    """Return the lines of one ticket. Line two onward carry two spaces."""
    head = "- [%s] **Ticket %s: %s.**" % (
        STATE_MARKS[ticket["claimed_state"]], ticket["id"], ticket["title"])
    tail = ("%s %s" % (CRITERIA_MARKER, ticket["criteria"])).rstrip()
    body = ticket["body"].strip()
    parts = body.split("\n") if body else [""]
    parts[-1] = ("%s %s" % (parts[-1], tail)).strip()
    lines = [("%s %s" % (head, parts[0])).rstrip()]
    for part in parts[1:]:
        lines.append(CONTINUATION_PREFIX + part.strip())
    return lines

File: src/plan/plan.ts
/**
 * The plan document format. A verbatim port of prototype/aidos_proto/plan.py.
 * Pure functions. PLAN.md / SPEC.md section 12 is the contract.
 * If you change this parser, also update skills/aidos-plan/verify-plan.mjs, which mirrors it (and vice versa): keep the two in sync. */

import matter from "gray-matter";
import YAML from "yaml";

import type { ContextSection, TicketState } from "../kernel/types";
import { PlanParseError } from "../kernel/types";

/** The mark that each line carries, and the state the mark claims. */
export const MARK_STATES: Record<string, TicketState> = {
  " ": "open",
  "~": "in_progress",
  "?": "awaiting_verification",
  x: "done",
};

/** The reverse map, which the renderer uses. */
export const STATE_MARKS: Record<TicketState, string> = {
  open: " ",
  in_progress: "~",
  awaiting_verification: "?",
  done: "x",
};

// The fence that opens and closes the frontmatter.
const FENCE = "---";

// The prefix of every heading that opens a phase or a context section.
const HEADING_PREFIX = "## ";

// A heading that opens a phase. Every other `## ` heading opens a context
// section.
const PHASE_HEADING = /^## Phase (\d+): (.+)$/;

// The state suffix of a phase title: an em dash and one backticked state
// word. The suffix is display text, not part of the title.
const PHASE_STATE_SUFFIX = / — `[^`]+`$/;

// The prefix of a continuation line.
const CONTINUATION_PREFIX = "  ";

// The marker that separates the body of a ticket from its criteria.
const CRITERIA_MARKER = "**Evaluate:**";
// M6 fix: title may contain dots; the split between title and body is the
// last ".\*\*" before body, not the first dot. Capture title greedily up to
// the final ".\*\*" on the line; verify by requiring the trailing ".\*\*".
const TICKET_LINE = /^- \[([ ~?x])\] \*\*Ticket ([^:]+): (.+)\.\*\*\s?(.*)$/;

/**
 * One phase heading of the document, deduped, in document order. The raw
 * text holds the heading after the `## ` prefix, so the renderer re-emits
 * the heading with one prefix.
 */
export interface PlanPhase {
  number: number;
  title: string;
  raw: string;
}

export interface PlanTicket {
  id: string;
  title: string;
  body: string;
  criteria: string;
  claimedState: TicketState;
  order: number;
  phase: number;
}

export interface PlanDocument {
  frontmatter: string;
  frontmatterData: Record<string, unknown>;
  preamble: string;
  contextSections: ContextSection[];
  tickets: PlanTicket[];
  phases: PlanPhase[];
}

/** The working data of one ticket before its body is complete. */
interface RawTicket {
  id: string;
  title: string;
  lines: string[];
  claimedState: TicketState;
  order: number;
  line: number;
  phase: number;
}

/** Read one plan document. Throws PlanParseError on the first bad line. */
export function parsePlan(text: string): PlanDocument {
  const lines = text.split("\n");
  const frontmatter = _takeFrontmatter(lines);
  const preamble = _takePreamble(lines, frontmatter.index);
  const contextSections: ContextSection[] = [];
  const rawTickets: RawTicket[] = [];
  // Tickets before the first phase heading take phase 1. A heading that is
  // not a phase marker never resets the phase.
  let currentPhase = 1;
  const phases: PlanPhase[] = [];
  const phaseNumbers = new Set<number>();
  let index = preamble.index;
  while (index < lines.length) {
    const line = lines[index];
    if (line.trim() === "") {
      index += 1;
      continue;
    }
    if (_isHeading(line)) {
      const phase = _phaseOfHeading(line);
      if (phase) {
        // A phase marker sets the phase of every ticket after it. The first
        // heading of one number is the record the renderer re-emits. The
        // prose between the marker and the next heading or ticket is
        // consumed and dropped, like context-section prose.
        currentPhase = phase.number;
        if (!phaseNumbers.has(phase.number)) {
          phaseNumbers.add(phase.number);
          phases.push(phase);
        }
        index += 1;
        while (
          index < lines.length &&
          !_isHeading(lines[index]) &&
          !TICKET_LINE.test(lines[index])
        ) {
          index += 1;
        }
        continue;
      }
      const section = _takeContextSection(lines, index, contextSections.length);
      contextSections.push(section.section);
      index = section.index;
      continue;
    }
    if (line.startsWith(CONTINUATION_PREFIX)) {
      if (rawTickets.length === 0) {
        throw new PlanParseError(
          index + 1,
          `line ${index + 1} continues a ticket, but the document holds no ticket yet`,
        );
      }
      rawTickets[rawTickets.length - 1].lines.push(line.trim());
      index += 1;
      continue;
    }
    const ticketMatch = TICKET_LINE.exec(line);
    if (!ticketMatch) {
      throw new PlanParseError(
        index + 1,
        `line ${index + 1} is neither a ticket line nor a continuation line`,
      );
    }
    rawTickets.push(
      _startTicket(ticketMatch, index + 1, rawTickets.length + 1, currentPhase),
    );
    index += 1;
  }
  return {
    frontmatter: frontmatter.text,
    frontmatterData: _parseFrontmatterData(frontmatter.text),
    preamble: preamble.text,
    contextSections,
    phases,
    tickets: rawTickets.map((raw) => _finishTicket(raw)),
  };
}

/**
 * Write one plan document. The tickets keep their given order. The phase
 * headings emit in phase-number order, and the tickets of a phase with no
 * record emit before every phase heading.
 */
export function renderPlan(doc: PlanDocument): string {
  const blocks: string[] = [];
  const frontmatter = doc.frontmatter.trim();
  if (frontmatter) {
    blocks.push(frontmatter);
  }
  const preamble = doc.preamble.trim();
  if (preamble) {
    blocks.push(preamble);
  }
  for (const section of doc.contextSections) {
    blocks.push(_renderContextSection(section));
  }
  for (const group of _ticketGroupsOf(doc)) {
    if (group.phase) {
      blocks.push(HEADING_PREFIX + group.phase.raw);
    }
    for (const ticket of group.tickets) {
      blocks.push(_renderTicket(ticket).join("\n"));
    }
  }
  if (blocks.length === 0) {
    return "";
  }
  return blocks.join("\n\n") + "\n";
}

/**
 * Group the tickets for rendering: first the tickets of a phase with no
 * phase record, then one group per recorded phase, ascending by number.
 */
function _ticketGroupsOf(
  doc: PlanDocument,
): Array<{ phase?: PlanPhase; tickets: PlanTicket[] }> {
  const recorded = [...doc.phases].sort((a, b) => a.number - b.number);
  const groups: Array<{ phase?: PlanPhase; tickets: PlanTicket[] }> = [];
  const unrecorded = doc.tickets.filter(
    (ticket) => !recorded.some((phase) => phase.number === ticket.phase),
  );
  if (unrecorded.length > 0) {
    groups.push({ tickets: unrecorded });
  }
  for (const phase of recorded) {
    groups.push({
      phase,
      tickets: doc.tickets.filter((ticket) => ticket.phase === phase.number),
    });
  }
  return groups;
}

// ---- reading ----

/** Take the frontmatter text and the index of the line after it. */
function _takeFrontmatter(
  lines: readonly string[],
): { text: string; index: number } {
  if (lines.length === 0 || lines[0].trim() !== FENCE) {
    return { text: "", index: 0 };
  }
  for (let index = 1; index < lines.length; index++) {
    if (lines[index].trim() === FENCE) {
      return { text: lines.slice(0, index + 1).join("\n"), index: index + 1 };
    }
  }
  throw new PlanParseError(1, "the frontmatter never closes");
}

/** Parse the raw frontmatter block with gray-matter and the yaml engine. */
function _parseFrontmatterData(raw: string): Record<string, unknown> {
  if (raw === "") {
    return {};
  }
  let parsed: unknown;
  try {
    parsed = matter(raw, {
      engines: {
        yaml: {
          parse: (s) => YAML.parse(s),
          stringify: (o) => YAML.stringify(o),
        },
      },
    }).data;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new PlanParseError(1, "the frontmatter is not valid YAML: " + message);
  }
  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
    return {};
  }
  return parsed as Record<string, unknown>;
}

/** The preamble text and the index of the first heading or ticket. */
function _takePreamble(
  lines: readonly string[],
  start: number,
): { text: string; index: number } {
  let index = start;
  while (
    index < lines.length &&
    !_isHeading(lines[index]) &&
    !TICKET_LINE.test(lines[index])
  ) {
    index += 1;
  }
  return { text: _trimBlankLines(lines.slice(start, index)), index };
}


/** One context section and the index of the line that ends it. */
function _takeContextSection(
  lines: readonly string[],
  start: number,
  sectionNumber: number,
): { section: ContextSection; index: number } {
  let index = start + 1;
  while (
    index < lines.length &&
    !_isHeading(lines[index]) &&
    !TICKET_LINE.test(lines[index])
  ) {
    index += 1;
  }
  return {
    section: {
      heading: lines[start].replace(/\s+$/, ""),
      text: _trimBlankLines(lines.slice(start + 1, index)),
      index: sectionNumber,
    },
    index,
  };
}

/** The working data of one ticket, before its body is complete. */
function _startTicket(
  match: RegExpExecArray,
  lineNumber: number,
  order: number,
  phase: number,
): RawTicket {
  const first = match[4].trim();
  return {
    id: match[2],
    title: match[3],
    lines: first ? [first] : [],
    claimedState: MARK_STATES[match[1]],
    order,
    line: lineNumber,
    phase,
  };
}

/** Split the body of one ticket from its criteria. */
function _finishTicket(raw: RawTicket): PlanTicket {
  const text = raw.lines.join("\n");
  if (!text.includes(CRITERIA_MARKER)) {
    throw new PlanParseError(
      raw.line,
      `line ${raw.line} starts a ticket that holds no ${CRITERIA_MARKER} marker`,
    );
  }
  // The marker sits alone on its own continuation line. A marker line with
  // text after the marker is the old format, and the old format is refused.
  const markerLineAt = raw.lines.findIndex((line) =>
    line.includes(CRITERIA_MARKER),
  );
  const markerLine = raw.lines[markerLineAt];
  if (markerLine !== CRITERIA_MARKER) {
    throw new PlanParseError(
      raw.line,
      `line ${raw.line} holds a ${CRITERIA_MARKER} marker with text on the same line. Put the marker alone on its own line, and put "- criterion" lines after it.`,
    );
  }
  // Blank lines are dropped at read time, so every line after the marker is
  // one list item, or one wrap of the list item before it.
  const criteriaLines = raw.lines.slice(markerLineAt + 1).filter(
    (line) => line !== "",
  );
  if (criteriaLines.length === 0 || !criteriaLines[0].startsWith("- ")) {
    throw new PlanParseError(
      raw.line,
      `line ${raw.line} holds a ${CRITERIA_MARKER} marker with no list. Put one "- criterion" line after the marker for each criterion.`,
    );
  }
  const criteria: string[] = [];
  for (const line of criteriaLines) {
    if (line.startsWith("- ")) {
      criteria.push(line.slice(2).trim());
    } else {
      // A line with no "- " mark wraps the criterion before it. Join the
      // wrap with one space.
      criteria[criteria.length - 1] += " " + line.trim();
    }
  }
  return {
    id: raw.id,
    title: raw.title,
    body: raw.lines.slice(0, markerLineAt).join("\n").trim(),
    criteria: criteria.join("\n"),
    claimedState: raw.claimedState,
    order: raw.order,
    phase: raw.phase,
  };
}

/** Say whether one line opens a phase or a context section. */
function _isHeading(line: string): boolean {
  return line.startsWith(HEADING_PREFIX);
}

/** The phase one heading opens, or undefined when the heading is not one. */
function _phaseOfHeading(line: string): PlanPhase | undefined {
  const match = PHASE_HEADING.exec(line);
  if (!match) {
    return undefined;
  }
  return {
    number: Number(match[1]),
    title: _stripStateSuffix(match[2]),
    raw: line.slice(HEADING_PREFIX.length).replace(/\s+$/, ""),
  };
}

/** Drop the em-dash state suffix of a phase title, when present. */
function _stripStateSuffix(title: string): string {
  return title.replace(PHASE_STATE_SUFFIX, "").trim();
}

/** Join lines, and drop the blank lines at the start and at the end. */
function _trimBlankLines(block: readonly string[]): string {
  let start = 0;
  let end = block.length;
  while (start < end && block[start].trim() === "") {
    start += 1;
  }
  while (end > start && block[end - 1].trim() === "") {
    end -= 1;
  }
  return block.slice(start, end).join("\n");
}

// ---- writing ----

/** The text of one context section, heading first. */
function _renderContextSection(section: ContextSection): string {
  const text = section.text.trim();
  if (!text) {
    return section.heading;
  }
  return `${section.heading}\n\n${text}`;
}


/**
 * The lines of one ticket. The body lines after the first carry two spaces.
 * The marker and every criteria line carry two spaces, and a blank line sits
 * between the body, the marker, and the criteria list. The document then
 * re-imports to the same fields.
 */
function _renderTicket(ticket: PlanTicket): string[] {
  const head = `- [${STATE_MARKS[ticket.claimedState]}] **Ticket ${ticket.id}: ${ticket.title}.**`;
  const body = ticket.body.trim();
  const bodyLines = body ? body.split("\n") : [];
  const lines = [`${head} ${bodyLines[0] ?? ""}`.trimEnd()];
  for (const part of bodyLines.slice(1)) {
    lines.push(CONTINUATION_PREFIX + part.trim());
  }
  lines.push("");
  lines.push(CONTINUATION_PREFIX + CRITERIA_MARKER);
  lines.push("");
  for (const criterion of ticket.criteria.trim().split("\n")) {
    lines.push(`${CONTINUATION_PREFIX}- ${criterion.trim()}`);
  }
  return lines;
}

export { PlanParseError };
